File: internal/model/deadletter.go
package model

// DeadLetterQueue is a queue that receives what other queues could not keep.
//
// It is not the shape DeadLetterReader describes, and that is the point.
// RocketMQ gives every consumer group a dead-letter topic of its own, named
// after it, so a group name is enough to find one. RabbitMQ has no such thing:
// a queue is declared with a dead-letter exchange, that exchange routes like
// any other, and whatever it routes to becomes a dead-letter queue by
// convention rather than by declaration. Finding one means walking the
// topology backwards.
type DeadLetterQueue struct {
	Namespace string `json:"namespace"`
	// Name is the queue dead letters land in. It has no special status on the
	// broker - it is an ordinary queue that something else points at.
	Name  string `json:"name"`
	Depth int64  `json:"depth"`
	// Consumers is how many are draining it. A dead-letter queue with a
	// consumer is a retry pipeline; one without is a backlog nobody is
	// looking at, which is the case worth surfacing.
	Consumers int `json:"consumers"`

	// Sources are the queues whose dead-letter exchange routes here, and the
	// reason this queue matters. A dead-letter queue with no sources is one
	// whose producers were deleted or reconfigured, and it will never receive
	// anything again.
	Sources []*DeadLetterSource `json:"sources"`
}

// DeadLetterSource is one queue that dead-letters into another.
type DeadLetterSource struct {
	Queue string `json:"queue"`
	// Subscription is which reader of the source queue dead-lettered, where
	// the family attaches the policy to a subscriber rather than to the queue.
	// A Pulsar dead-letter topic is named "<topic>-<subscription>-DLQ", so
	// without this the page could name the topic and not who gave up on it -
	// and one topic read by five subscriptions has five separate answers.
	//
	// Empty on RabbitMQ, where the policy belongs to the queue itself.
	Subscription string `json:"subscription"`
	// Exchange is what the source queue was declared to dead-letter through.
	Exchange string `json:"exchange"`
	// RoutingKey is the key the message is re-published with. Empty means the
	// message keeps its original routing key, which is the default and changes
	// where it lands.
	RoutingKey string `json:"routingKey"`
}

File: frontend/src/mq/pulsar/deadletter.ts
/**
 * Pulsar's view of the canonical dead-letter model.
 *
 * There is no broker-side dead-letter object on this family. What there is, is
 * a convention in the official client libraries: a consumer with a DLQ policy
 * republishes to "<topic>-<subscription>-DLQ", and retries go to
 * "<topic>-<subscription>-RETRY". Both are ordinary topics that happen to be
 * named that way, which is why the driver finds them by walking a namespace
 * and why a row can exist with no source at all.
 */
import type { DeadLetterQueue } from "@bindings/model/models";

/** Which half of the convention a topic follows. */
export const DeadLetterKind = { Dlq: "dlq", Retry: "retry" } as const;
export type DeadLetterKindValue = (typeof DeadLetterKind)[keyof typeof DeadLetterKind];

/**
 * A retry topic is a pipeline; a DLQ is where it ends up.
 *
 * Worth distinguishing on the page: a growing retry topic means consumers are
 * failing and recovering, and a growing DLQ means they have given up.
 */
export function kindOf(queue: DeadLetterQueue): DeadLetterKindValue {
  return queue.name.endsWith("-RETRY") ? DeadLetterKind.Retry : DeadLetterKind.Dlq;
}

/** The topic that dead-lettered here, or "" when its origin is gone. */
export function sourceTopic(queue: DeadLetterQueue): string {
  return queue.sources?.[0]?.queue ?? "";
}

/**
 * The subscription that gave up, or "".
 *
 * This is the answer the page exists for: one topic read by five
 * subscriptions has five separate dead-letter topics, and naming only the
 * topic would not say which reader is failing.
 */
export function sourceSubscription(queue: DeadLetterQueue): string {
  return queue.sources?.[0]?.subscription ?? "";
}

/**
 * Whether this topic's origin could not be resolved.
 *
 * An orphan holds a backlog nothing will ever drain and nobody will ever look
 * at, which is the single most useful row on the page - so it is drawn as a
 * finding rather than as a row with blank columns.
 */
export const isOrphaned = (queue: DeadLetterQueue): boolean =>
  (queue.sources?.length ?? 0) === 0;

/** A figure the driver did not report, as null rather than zero. */
export function reported(value: number): number | null {
  return value === -1 ? null : value;
}
